### PR TITLE
feat: use self-signed JWTs if alwaysUseJWTAccessWithScope is true

### PR DESCRIPTION
**Pull request by sofisl**
Work for the self-signed JWT implementation
In conjunction with: https://github.com/googleapis/gapic-generator-typescript/pull/908

### CODE REVIEW

**Comment by sofisl** [comment]
Still need to add tests

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,4 @@ package-lock.json
 yarn.lock
 dist/
 __pycache__
+.DS_Store
```
Comment (JustinBeckwith) [inline .gitignore:16]: Please add a newline to make the (-) go away

File: src/auth/googleauth.ts [hunk 2]
```diff
@@ -111,6 +111,8 @@ export class GoogleAuth {
    * @private
    */
   private checkIsGCE?: boolean = undefined;
+  useJWTAccessAlways?: boolean;
```
Comment (JustinBeckwith) [inline src/auth/googleauth.ts:114]: Can we bike shed this name a little bit?  From a customer perspective, I'm not sure if I'd know what `useJWTAccessAlways` would mean.  Would it make more sense to use something like `useSelfSignedJWT`?

File: src/auth/googleauth.ts [hunk 3]
```diff
@@ -111,6 +111,8 @@ export class GoogleAuth {
...
+  defaultServicePath: string | undefined;
```
Comment (JustinBeckwith) [inline src/auth/googleauth.ts:115]: If something is `type | undefined`, we typically use the `?` like we do elsewhere

File: src/auth/jwtclient.ts [hunk 4]
```diff
@@ -26,13 +26,15 @@ import {
 } from './oauth2client';
 
 export interface JWTOptions extends RefreshOptions {
+  useJWTAccessAlways?: boolean;
```
Comment (JustinBeckwith) [inline src/auth/jwtclient.ts:29]: It would be nice to add descriptions to these properties so they show up in the documentation.  I know we don't have any existing ones here yet :/

File: src/auth/jwtclient.ts [hunk 5]
```diff
@@ -68,7 +71,9 @@ export class JWT extends OAuth2Client implements IdTokenProvider {
     key?: string,
     scopes?: string | string[],
     subject?: string,
-    keyId?: string
+    keyId?: string,
+    useJWTAccessAlways?: boolean,
```
Comment (JustinBeckwith) [inline src/auth/jwtclient.ts:75]: Let's just stop adding new variables to the constructor, and if they want this, force folks to use the object approach.  This was largely here for back-compat.

File: src/auth/googleauth.ts [hunk 6]
```diff
@@ -456,6 +458,8 @@ export class GoogleAuth {
     } else {
       (options as JWTOptions).scopes = this.scopes;
       client = new JWT(options);
+      client.defaultServicePath = this.defaultServicePath;
+      client.useJWTAccessAlways = this.useJWTAccessAlways || false;
```
Comment (JustinBeckwith) [inline src/auth/googleauth.ts:462]: You shouldn't need to include the ` || false` here.  The `||` checks if the variable has a falsey value, which would get evaluated to `false` anyways.

Comment (JustinBeckwith) [reply src/auth/googleauth.ts:462]: here and below

File: src/auth/jwtclient.ts [hunk 7]
```diff
@@ -121,44 +128,63 @@ export class JWT extends OAuth2Client implements IdTokenProvider {
...
+      if (!this.access) {
+        this.access = new JWTAccess(
+          this.email,
+          this.key,
+          this.keyId,
+          this.eagerRefreshThresholdMillis
+        );
+      }
+
+      const audience = 'https://' + this.defaultServicePath + '/';
```
Comment (JustinBeckwith) [inline src/auth/jwtclient.ts:141]: Please use string interpolation

File: src/auth/googleauth.ts (hunk 2 above)
Comment (bcoe) [reply src/auth/googleauth.ts:114]: @JustinBeckwith I believe this to be consistent with other language's implementation.

My recommendation was that we allow this as a class variable, but don't advertise it as an option on the options interface. The reason being that we want to be able to set this option in GAPIC libraries, but don't want to encourage customers to set the variable.

File: src/auth/jwtclient.ts (hunk 4 above)
Comment (bcoe) [reply src/auth/jwtclient.ts:29]: I'd like to advocate that we don;'t advertise this option in any interfaces, and just set it on the class.

This is what we did for `defaultScopes`.

File: src/auth/googleauth.ts (hunk 2 above)
Comment (JustinBeckwith) [reply src/auth/googleauth.ts:114]: okies - my feedback ain't blocking, and consistency is good

File: test/test.googleauth.ts [hunk 8]
```diff
@@ -402,6 +402,20 @@ describe('googleauth', () => {
       assert.strictEqual(json.private_key, (result as JWT).key);
     });
 
+    it.only('fromJSON should set default access variable with private key', () => {
```
Comment (JustinBeckwith) [inline test/test.googleauth.ts:405]: blocking so we don't merge with the `.only`

**Review by JustinBeckwith — CHANGES_REQUESTED** [review]

File: test/test.googleauth.ts (hunk 8 above)
Comment (sofisl) [reply test/test.googleauth.ts:405]: yep, sorry still haven't pushed the new changes!

**Comment by sofisl** [comment]
@bcoe @JustinBeckwith, comments  have been addressed & tests added!

File: src/auth/googleauth.ts [hunk 9]
```diff
@@ -580,6 +580,8 @@ export class GoogleAuth {
                 ...this.clientOptions,
                 keyFile: this.keyFilename,
               });
+              client.defaultServicePath = this.defaultServicePath;
```
Comment (bcoe) [inline src/auth/googleauth.ts:583]: we should need this because we have the helper `client.setGapicJWTValues`.

File: src/auth/googleauth.ts [hunk 10]
```diff
@@ -600,6 +602,8 @@ export class GoogleAuth {
   fromAPIKey(apiKey: string, options?: RefreshOptions): JWT {
     options = options || {};
     const client = new JWT(options);
+    client.defaultServicePath = this.defaultServicePath;
```
Comment (bcoe) [inline src/auth/googleauth.ts:605]: I think here you just want to call `setGapicJWTValues()`.

**Review by bcoe — COMMENTED** [review]
LGTM, thank you for doing this.

**Comment by sofisl** [comment]
Waiting for: https://github.com/googleapis/gapic-generator-typescript/pull/908

**Review by bshaffer — COMMENTED** [review]
Same feedback here as in the gapic generator: I'd advocate for all of these to be renamed from `useJWTAccessAlways` to `useJwtAccessWithScope` for consistency (see [AIP #4111](https://google.aip.dev/auth/4111#scope-vs-audience)), unless this was a conscious decision with justification.

**Comment by bshaffer** [comment]
> Same feedback here as in the gapic generator: I'd advocate for all of these to be renamed from `useJWTAccessAlways` to `useJwtAccessWithScope` for consistency (see [AIP #4111](https://google.aip.dev/auth/4111#scope-vs-audience)), unless this was a conscious decision with justification.

As https://github.com/googleapis/google-auth-library-nodejs/pull/1204 is already released and uses the name `useJWTAccessAlways`, it's ok to keep that variable name here, and this should not block rolling the feature out!

File: src/auth/jwtclient.ts [hunk 11]
```diff
@@ -122,44 +122,63 @@ export class JWT extends OAuth2Client implements IdTokenProvider {
   protected async getRequestMetadataAsync(
     url?: string | null
   ): Promise<RequestMetadataResponse> {
-    if (!this.apiKey && !this.hasUserScopes() && url) {
-      if (
-        this.additionalClaims &&
-        (
-          this.additionalClaims as {
-            target_audience: string;
+    if (this.useJWTAccessAlways) {
```
Comment (bshaffer) [inline src/auth/jwtclient.ts:125]: This is interesting logic here because it's saying that for clients which set `useJWTAccessAlways` to `true`, do not allow API Key auth or ID token auth. I'm not sure this is what we want to go so far as to say that. The more accurate way (I think) would be something like this:

```js
var canSelfSign = (this.hasUserScopes() && this.useJWTAccessAlways) // self-signed JWT with scopes
  || (this.hasDefaultScopes() && url) // self-signed JWT with audience
if (!this.apiKey && canSelfSign) {
```

Where it used to have:

```js
if (!this.apiKey && !this.hasUserScopes() && url) {
```

Comment (sofisl) [reply src/auth/jwtclient.ts]: @bshaffer, I gave this a try earlier today and tests fail, both the ones we've added and the current functionality.

I think this is because we check later if we should sign with audience or scopes, in `jwtaccess.ts`, line 99:

```
if (scopes) {
      defaultClaims = {
        iss: this.email,
        sub: this.email,
        scope: scopes,
        exp,
        iat,
      };
    } else {
      // If either scopes are not specified, or if audience is something unique,
      // just use default service endpoint.
      defaultClaims = {
        iss: this.email,
        sub: this.email,
        aud: url,
        exp,
        iat,
      };
    }
```

> This is interesting logic here because it's saying that for clients which set useJWTAccessAlways to true, do not allow API Key auth or ID token auth

I'm curious about this, isn't a part of this RFC to always use self-signed JWT if it's coming from a GAPIC client?

**Review by bcoe — APPROVED** [review]

**Review by JustinBeckwith — APPROVED** [review]

**Review by bcoe — CHANGES_REQUESTED** [review]
I believe @bshaffer is correct about the `authToken` edge case.

**Review by bcoe — APPROVED** [review]
To me this is looking really good, and I think the tests are comprehensive.

It would be good to have Brent do one more sniff test.

**Review by bcoe — CHANGES_REQUESTED** [review]
we have a final session planned Monday to get this over finish line.

File: test/test.jwt.ts [hunk 12]
```diff
@@ -22,6 +22,7 @@ import * as sinon from 'sinon';
 import {GoogleAuth, JWT} from '../src';
 import {CredentialRequest, JWTInput} from '../src/auth/credentials';
 import * as jwtaccess from '../src/auth/jwtaccess';
+import {stub} from 'sinon';
```
Comment (bcoe) [inline test/test.jwt.ts:25]:
```suggestion
```

**Review by bcoe — APPROVED** [review]

File: src/auth/jwtclient.ts [hunk 13]
```diff
@@ -148,10 +152,22 @@ export class JWT extends OAuth2Client implements IdTokenProvider {
...
+        if (this.hasUserScopes()) {
+          scopes = this.scopes;
+        } else if (!url) {
+          scopes = this.defaultScopes;
+        }
+
         const headers = await this.access.getRequestHeaders(
-          url,
-          this.additionalClaims
+          url ?? undefined,
```
Comment (bshaffer) [inline src/auth/jwtclient.ts:164]: NodeJS question: why isn't it enough to just pass in the falsy value (e.g. `url` without the `?? undefined` check)? Is it for cases where it's `false` or an empty string?

**Review by bshaffer — APPROVED** [review]
One question, otherwise this looks amazing!

File: src/auth/jwtclient.ts (hunk 13 above)
Comment (bcoe) [reply src/auth/jwtclient.ts:164]: `null` and `false` are considered valid arguments for an optional string parameter, which can only be a string or `undefined`.
